File: leetcode/1001-1999/1930-unique-length-3-palindromic-subsequences/UniqueLength3PalindromicSubsequences.java
import java.util.HashSet;
import java.util.Set;

public class UniqueLength3PalindromicSubsequences {
  public static void main(String[] args) {
    System.out.println(new Solution().countPalindromicSubsequence("bbcbaba"));
  }
}

class Solution {
  public int countPalindromicSubsequence(String s) {
    int total = 0;

    for (char c = 'a'; c <= 'z'; c++) {
      int left = s.indexOf(c);
      int right = s.lastIndexOf(c);

      if (right > left) {
        Set<Character> middles = new HashSet<>();

        for (int i = left + 1; i < right && middles.size() < 26; i++) {
          middles.add(s.charAt(i));
        }

        total += middles.size();
      }
    }

    return total;
  }
}

File: leetcode/1001-1999/1930-unique-length-3-palindromic-subsequences/UniqueLength3PalindromicSubsequences.ts
function countPalindromicSubsequence(s: string): number {
  let total = 0;

  for (let i = 0; i < 26; i++) {
    let c = String.fromCharCode(97 + i);

    let left = s.indexOf(c);
    let right = s.lastIndexOf(c);

    let characters = new Set();

    for (let i = left + 1; i < right; i++) {
      characters.add(s[i]);
    }

    total += characters.size;
  }

  return total;
};

console.log(countPalindromicSubsequence("aabca"));
